### ISSUE TITLE
Error thrown by stateless component while class component doesn't

### ISSUE THREAD

**Issue by CodKnight** [facebook/flow#3195]
Hi all, 
I am encountering an issue I don't understand, I don't even know if it is an issue or a normal behavior. 
 
```
/* @flow */
import React from 'react';

type ComponentsProps = {
  name: string,
}

type AppProps = {
  name?: string,
}

const StatelessComponent = ({ name }: ComponentsProps) => (
  <div>{name}</div>
);

class ClassComponent extends React.Component {
  props: ComponentsProps

  render() {
    const { name } = this.props;

    return (<div>{name}</div>)
  }
}

class PureClassComponent extends React.PureComponent {
  props: ComponentsProps

  render() {
    const { name } = this.props;

    return (<div>{name}</div>)
  }
}

const App = ({ name }: AppProps) =>  {
  return (
    <div className="App">
      <StatelessComponent name={name} />
      <ClassComponent name={name} />
      <PureClassComponent name={name} />
    </div>
  );
}
```

This code displays an error : 
![screenshot from 2017-01-16 12-12-20](https://cloud.githubusercontent.com/assets/10506209/21980873/1424fe74-dbe5-11e6-8046-dda8715d9c60.png)

I understand the error and it seems perfectly logical to me. 
But the error is only thrown on the stateless component, even if those 3 components are exactly the same except the fact the 2 others are not stateless.
You can see the entire project [here](https://github.com/CodKnight/flowissue-example).

Is there something I am missing or is it a bug ?

**Comment by AugustinLF** [comment]
In AppProps, you write that name might not exist (`name?: string`), which doesn't match the expectations by StatelessComponent (which requires a string). You could fix this example by [passing a default value](https://flowtype.org/try/#0PQKgBAAgZgNg9gdzCYAoAlgWwA5wE4AuYASgKYCGAxkVHnJmAOR4XWMDcqqBAntqWADC9XADtSoggGcACnWxSwAXjABvVGDCjymUgC4wUgnnSiA5gBpUAXy69+YAILZscuAuVqNWnaQD8BkYm5la2qJRwokZgAMoE5ASkMKRSUsI4kRJEKgAUqj66YNYG6WJZsvJSAJTKAHxgOd4APAAm6AButarautZNwG2dqFWc4TDkqULjqaWZkmCkAB6Joi2KZFQEAHSz4vPqmtiVJSJz0m4KXJosq6R4OTUHmmARUUT5PQLWngQAFuhSLZHdxSUbPMAsAgAVzwogarQ6XU+fQGiKq3lsYUo00UMhhpEEON2WQWywkaxIrG2eJYxP23mBChOGT250qVwh5LuDy84Ne0Q+viKP3+gMZoI511I0Nh8MGSN8KPl6M0mK4-KIzmwnjyBS+Bi1F2qdU0T0hMLhjXBCPaLxxADlfEoAERa521bzWuIJJIpNKnVl6pTdIUAH1DTEY32AHvBmiahIm-pZJM+weRYBjnueTRpBKJAdTTpDvUzsZzqKGmhGNiAA) for name to StatelessComponent.

`<StatelessComponent name={name || ''} />`

**Comment by CodKnight** [comment]
@AugustinLF 
Yes, I spotted my mistake and I understand it. That's not what I'm asking. I wrote the ClassComponent first, and had no error so I didn't spotted the mistake. I only spotted it when I reuse those types with a stateless component. If I had not had to write a stateless, I probably would have missed it

**Comment by asolove** [comment]
This is now fixed in Flow 0.53. If I make slight syntax updates for that version, I now get the expected type errors for all three component types.

```js
/* @flow */
import React from 'react';

type ComponentsProps = {
  name: string,
}

type AppProps = {
  name?: string,
}

const StatelessComponent = ({ name }: ComponentsProps) => (
  <div>{name}</div>
);

class ClassComponent extends React.Component<ComponentsProps, *> {

  render() {
    const { name } = this.props;

    return (<div>{name}</div>)
  }
}

class PureClassComponent extends React.PureComponent<ComponentsProps, *> {
  props: ComponentsProps

  render() {
    const { name } = this.props;

    return (<div>{name}</div>)
  }
}

const App = ({ name }: AppProps) =>  {
  return (
    <div className="App">
      <StatelessComponent name={name} />
      <ClassComponent name={name} />
      <PureClassComponent name={name} />
    </div>
  );
}
```
